### PR TITLE
gh-91054: Add code object watchers API

### PR DESCRIPTION
**Pull request by itamaro**
This PR implements gh-91054, allowing extensions to set a callback function to be notified on creation and destruction of code objects

pyperformance results are 1.00x slower (~neutral) using AWS bare metal machine ([full comparison gist](https://gist.github.com/itamaro/fbcf5733d40fc3957d952f54d6d3e9e9))

Co-authored-by: @Ye11ow-Flash

### CODE REVIEW

**Comment by carljm** [comment]
@itamaro I talked this over briefly with @markshannon and it sounds like we should make the same optimization here as @mpage did in https://github.com/python/cpython/pull/98175/commits/812dd5fa42ad2f3d6820838a4e18ea9dfc94bff2 to make the check for active watchers a bit more efficient and reduce the overhead on code object creation

File: Objects/codeobject.c [hunk 1]
```diff
@@ -12,6 +12,62 @@
 #include "clinic/codeobject.c.h"
 
 
+static void
+notify_code_watchers(PyCodeEvent event, PyCodeObject *co)
```
Comment (markshannon) [inline Objects/codeobject.c:16]: Could you add a (8) bit vector of watchers, much as we did for function watchers.
That way, only one test is needed in the common case where there are no watchers.

Comment (itamaro) [reply Objects/codeobject.c:16]: sure, let me add it

**Comment by markshannon** [comment]
FYI, I expect to add a `PY_CODE_EVENT_INSTRUMENTED` event as part of PEP 669, when a code object gets instrumented by monitoring. I would expect tools to treat this new event the same as `PyFunction_EVENT_MODIFY_...`

File: Objects/codeobject.c [hunk 2]
```diff
@@ -12,6 +12,66 @@
...
+{
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    if (interp->active_code_watchers) {
```
Comment (markshannon) [inline Objects/codeobject.c:19]: I'm OK with this as is, but you could speed this up with an invariant.
`(active_code_watchers & (1 << i))` implies `interp->code_watchers[i] != NULL`

Then the loop can be something like this:
```
uint8_t watchers = interp->active_code_watchers;
while (watchers) {
    int watcher = left_most_bit(watchers);
    watchers &= ~(1 << watcher);
    if (interp->code_watchers[watcher](...) {
          PyErr_WriteUnraisable(...);
    }
}
```

This probably applies to the other watchers as well, so you might want to do this in another PR.

**Review by markshannon — APPROVED** [review]
Looks good. Does this complete the set of watchers?

File: Objects/codeobject.c (hunk 2 above)
Comment (carljm) [reply Objects/codeobject.c:19]: I think this optimization applies to func watchers, but is already effectively applied for dict and type watchers; in those cases we already never look up the watcher unless the bit is set.

**Comment by carljm** [comment]
> Looks good. Does this complete the set of watchers?

Yep! (At least the ones we had in mind.)

File: Objects/codeobject.c (hunk 2 above)
Comment (itamaro) [reply Objects/codeobject.c:19]: I have some more watchers-wide cleanup I wanted to do, so will add this optimization to the next PR, thanks!

**Comment by brettcannon** [comment]
FYI that buildbot failure doesn't seem to be transient. Are there new tests which are recursive and assuming a certain stack depth? Looking at the test failure it seems the call stack is being blown out in the emscripten build (line 515 of https://buildbot.python.org/all/#/builders/1050/builds/953/steps/10/logs/stdio).

**Comment by pablogsal** [comment]
As per our buildbot policy, this PR will be reverted if the buildbot fleet is not fixed in the next 24h.

**Comment by itamaro** [comment]
Thanks @pablogsal! I have opened GH-99978 to fix the issue. My local repro passes with the fix - could you get all buildbots to test this PR to confirm?

@brettcannon I couldn't find anything in this PR that could trigger deep recursion. I wonder if it's the same issue caught by the other buildbots, but manifesting in a different way on the emscripten buildbot?
do you have documentation on how to set up a local emscripten build in case the fix in GH-99978 is not sufficient?

**Comment by pablogsal** [comment]
> Thanks @pablogsal! I have opened [GH-99978](https://github.com/python/cpython/pull/99978) to fix the issue. My local repro passes with the fix - could you get all buildbots to test this PR to confirm?

Yup, checking!

File: Objects/codeobject.c (hunk 2 above)
Comment (itamaro) [reply Objects/codeobject.c:19]: cleanup and polish PR (including this optimization) is up at GH-99998

**Comment by brettcannon** [comment]
@itamaro https://github.com/python/cpython/tree/main/Tools/wasm covers how to do anything WebAssembly-related.

**Comment by itamaro** [comment]
thanks @brettcannon, sorry about the delay, I spent the last two days trying to repro and debug this 😬

**TLDR: I think this PR may have pushed something over some threshold that caused the junit xml generation to start hitting stack depth limits.** (but is not the root cause of the failures)

the steps in the docs failed (basically running `./Tools/wasm/wasm_build.py` in the docker container couldn't find the emsdk or something like that), and also didn't apply to the "node (pthreads)" target, which is the one failing.

I was able to repro the failure using some mix of the instructions and reverse engineering the buildbot itself.

TLDR build and test steps, inside the docker container:

```
mkdir -p build_oot/build build_oot/host
cd build_oot/build
../../configure && make -j
cd ../host
CONFIG_SITE=../../Tools/wasm/config.site-wasm32-emscripten emconfigure ../../configure --prefix '$(PWD)/target/host' --with-pydebug --without-pydebug --with-emscripten-target=node --disable-wasm-dynamic-linking --enable-wasm-pthreads --build=x86_64-pc-linux-gnu --host=wasm32-unknown-emscripten --with-build-python=$(pwd)/../build/python --build=$(../../config.guess) && emmake make -j2 all
make buildbottest 'TESTOPTS=-j2 --junit-xml test-results.xml -j2 ${BUILDBOT_TESTOPTS}' TESTPYTHONOPTS= TESTTIMEOUT=900
```

this is more or less what the buildbot does, and it fails (not always, but almost always).

this also runs the full test suite, but I was able to narrow it down to running just `test_capi` and still fail:

```
root@18f2ffc1b119:/python-wasm/cpython/build_oot/host# _PYTHON_HOSTRUNNER='/emsdk/node/14.18.2_64bit/bin/node --experimental-wasm-bigint --experimental-wasm-threads --experimental-wasm-bulk-memory' _PYTHON_PROJECT_BASE=/python-wasm/cpython/build_oot/host _PYTHON_HOST_PLATFORM=emscripten-wasm32 PYTHONPATH=/python-wasm/cpython/build_oot/host/build/lib.emscripten-wasm32-3.12:../../Lib _PYTHON_SYSCONFIGDATA_NAME=_sysconfigdata__emscripten_wasm32-emscripten /python-wasm/cpython/build_oot/host/../build/python  ../../Tools/scripts/run_tests.py --junit-xml test-results.xml test_capi
```

what makes me think this PR is not the root cause of the regression is that it consistently passes when running just the watchers tests:

```
root@18f2ffc1b119:/python-wasm/cpython/build_oot/host# _PYTHON_HOSTRUNNER='/emsdk/node/14.18.2_64bit/bin/node --experimental-wasm-bigint --experimental-wasm-threads --experimental-wasm-bulk-memory' _PYTHON_PROJECT_BASE=/python-wasm/cpython/build_oot/host _PYTHON_HOST_PLATFORM=emscripten-wasm32 PYTHONPATH=/python-wasm/cpython/build_oot/host/build/lib.emscripten-wasm32-3.12:../../Lib _PYTHON_SYSCONFIGDATA_NAME=_sysconfigdata__emscripten_wasm32-emscripten /python-wasm/cpython/build_oot/host/../build/python  ../../Tools/scripts/run_tests.py --junit-xml test-results.xml test_capi.test_watchers
```

and what makes me suspect the root cause is in junit xml generation is that dropping that flag causes test_capi to pass consistently:

```
root@18f2ffc1b119:/python-wasm/cpython/build_oot/host# _PYTHON_HOSTRUNNER='/emsdk/node/14.18.2_64bit/bin/node --experimental-wasm-bigint --experimental-wasm-threads --experimental-wasm-bulk-memory' _PYTHON_PROJECT_BASE=/python-wasm/cpython/build_oot/host _PYTHON_HOST_PLATFORM=emscripten-wasm32 PYTHONPATH=/python-wasm/cpython/build_oot/host/build/lib.emscripten-wasm32-3.12:../../Lib _PYTHON_SYSCONFIGDATA_NAME=_sysconfigdata__emscripten_wasm32-emscripten /python-wasm/cpython/build_oot/host/../build/python  ../../Tools/scripts/run_tests.py test_capi
```

I don't know enough about wasm and junit xml to dig deeper for a fix, but I hope this investigation can provide someone who's more familiar with wasm and junit xml with enough information to take it from here!

**Comment by brettcannon** [comment]
Thanks for the research! It looks like yesterday some change was made that brought back to being stable. But if it happens again I will definitely come back to your notes to see if we may need to tweak some code in how we're generating that junit XML output.
